DynamicPoolAllocator.hpp

```diff
@@ -59,32 +125,61 @@ class DynamicPoolAllocator
 
   // Allocate a new block and add it to the list of free blocks
   void allocateBlock(struct Block *&curr, struct Block *&prev, const std::size_t size) {
-    const std::size_t sizeToAlloc = std::max(alignmentAdjust(size), minBytes);
+    std::size_t sizeToAlloc;
+
+    if ( originalAllocations != NULL )
+      sizeToAlloc = std::max(alignmentAdjust(size), minBytes);
+    else
+      sizeToAlloc = std::max(alignmentAdjust(size), minInitialBytes);
+
     curr = prev = NULL;
     void *data = NULL;
 
     // Allocate data
-    data = allocator->allocate(sizeToAlloc);
+    try {
+      data = allocator->allocate(sizeToAlloc);
+    }
+    catch (...) {
+      freeOriginalAllocationBlocks();
+      data = allocator->allocate(sizeToAlloc);
+    }
+
     totalBytes += sizeToAlloc;
-    assert(data);
+    UMPIRE_ASSERT("Memory Allocation Failed" && data);
```

Comment:
This assert would never fail, since if `data` was not able to be allocated, the allocator would have already thrown an exception.

Reply:
It also relies on the UMPIRE_ASSERT macro...